API: DatetimeIndex union vs join different behavior with mismatched tzs

For union when we have mismatched tzs or mismatched tzawareness, we cast to object.

For join when we have mismatched tzawareness we raise TypeError, with mismatched tzs we cast to UTC.

It would be convenient if these had the same behavior.

One way to implement this would be to make `find_common_type` for mismatched dt64tzs return UTC.  Doing this would change the setop behavior in a pretty reasonable way.  the sticking point is in a couple of groupby tests:

```
    def test_transform_lambda_with_datetimetz():
        # GH 27496
        df = DataFrame(
            {
                "time": [
                    Timestamp("2010-07-15 03:14:45"),
                    Timestamp("2010-11-19 18:47:06"),
                ],
                "timezone": ["Etc/GMT+4", "US/Eastern"],
            }
        )
        result = df.groupby(["timezone"])["time"].transform(
            lambda x: x.dt.tz_localize(x.name)
        )
        expected = Series(
            [
                Timestamp("2010-07-15 03:14:45", tz="Etc/GMT+4"),
                Timestamp("2010-11-19 18:47:06", tz="US/Eastern"),
            ],
            name="time",
        )
>       tm.assert_series_equal(result, expected)
E       AssertionError: Attributes of Series are different
E       
E       Attribute "dtype" are different
E       [left]:  datetime64[ns, UTC]
E       [right]: object

    def test_groupby_multi_timezone(self):
    
        # combining multiple / different timezones yields UTC
    
        dates = [
            "2000-01-28 16:47:00",
            "2000-01-29 16:48:00",
            "2000-01-30 16:49:00",
            "2000-01-31 16:50:00",
            "2000-01-01 16:50:00",
        ]
        tzs = [
            "America/Chicago",
            "America/Chicago",
            "America/Los_Angeles",
            "America/Chicago",
            "America/New_York",
        ]
        df = DataFrame({"value": range(5), "date": dates, "tz": tzs})
    
        result = df.groupby("tz").date.apply(
            lambda x: pd.to_datetime(x).dt.tz_localize(x.name)
        )
    
        expected = Series(
            [
                Timestamp("2000-01-28 16:47:00-0600", tz="America/Chicago"),
                Timestamp("2000-01-29 16:48:00-0600", tz="America/Chicago"),
                Timestamp("2000-01-30 16:49:00-0800", tz="America/Los_Angeles"),
                Timestamp("2000-01-31 16:50:00-0600", tz="America/Chicago"),
                Timestamp("2000-01-01 16:50:00-0500", tz="America/New_York"),
            ],
            name="date",
            dtype=object,
        )
>       tm.assert_series_equal(result, expected)
E       AssertionError: Attributes of Series are different
E       
E       Attribute "dtype" are different
E       [left]:  datetime64[ns, UTC]
E       [right]: object
```

In both of these examples, its pretty reasonable to think the user would want the object-dtype result

--- a/doc/source/whatsnew/v1.3.0.rst
+++ b/doc/source/whatsnew/v1.3.0.rst
@@ -645,6 +645,7 @@ Deprecations
 - The ``inplace`` parameter of :meth:`Categorical.remove_categories`, :meth:`Categorical.add_categories`, :meth:`Categorical.reorder_categories`, :meth:`Categorical.rename_categories`, :meth:`Categorical.set_categories` is deprecated and will be removed in a future version (:issue:`37643`)
 - Deprecated :func:`merge` producing duplicated columns through the ``suffixes`` keyword  and already existing columns (:issue:`22818`)
 - Deprecated setting :attr:`Categorical._codes`, create a new :class:`Categorical` with the desired codes instead (:issue:`40606`)
+- Deprecated behavior of :meth:`DatetimeIndex.union` with mixed timezones; in a future version both will be cast to UTC instead of object dtype (:issue:`39328`)
 
 .. ---------------------------------------------------------------------------
 

--- a/pandas/core/indexes/base.py
+++ b/pandas/core/indexes/base.py
@@ -2930,6 +2930,21 @@ class Index(IndexOpsMixin, PandasObject):
                     "Can only union MultiIndex with MultiIndex or Index of tuples, "
                     "try mi.to_flat_index().union(other) instead."
                 )
+            if (
+                isinstance(self, ABCDatetimeIndex)
+                and isinstance(other, ABCDatetimeIndex)
+                and self.tz is not None
+                and other.tz is not None
+            ):
+                # GH#39328
+                warnings.warn(
+                    "In a future version, the union of DatetimeIndex objects "
+                    "with mismatched timezones will cast both to UTC instead of "
+                    "object dtype. To retain the old behavior, "
+                    "use `index.astype(object).union(other)`",
+                    FutureWarning,
+                    stacklevel=2,
+                )
 
             dtype = find_common_type([self.dtype, other.dtype])
             if self._is_numeric_dtype and other._is_numeric_dtype:

--- a/pandas/tests/indexes/datetimes/test_timezones.py
+++ b/pandas/tests/indexes/datetimes/test_timezones.py
@@ -1146,7 +1146,10 @@ class TestDatetimeIndexTimezones:
 
         rng2 = date_range("2012-11-15 12:00:00", periods=6, freq="H", tz="US/Eastern")
 
-        result = rng.union(rng2)
+        with tm.assert_produces_warning(FutureWarning):
+            # # GH#39328 will cast both to UTC
+            result = rng.union(rng2)
+
         expected = rng.astype("O").union(rng2.astype("O"))
         tm.assert_index_equal(result, expected)
         assert result[0].tz.zone == "US/Central"

